iOS/DuckDuckGo/AppLifecycle/AppStates/Terminating.swift

```diff
@@ -17,27 +17,110 @@
+/// It categorizes the error, reports it via a pixel, and either:
+/// - Terminates immediately, or
+/// - Shows a user-facing alert before termination (e.g. for disk space issues).
+///
+/// Unhandled errors result in a generic crash and pixel.
+///
 struct Terminating: TerminatingHandling {
 
-    let terminationError: UIApplication.TerminationError
+    private let application: UIApplication
 
-    init() {
+    init(error: Error, application: UIApplication = UIApplication.shared) {
         Logger.lifecycle.info("Terminating: \(#function)")
-        fatalError("App is in unrecoverable state")
-    }
 
-    init(terminationError: UIApplication.TerminationError,
-         application: UIApplication = UIApplication.shared) {
-        Logger.lifecycle.info("Terminating: \(#function)")
-        self.terminationError = terminationError
-        alertAndTerminate(application: application)
+        self.application = application
+
+        let mode: TerminationMode
+        let pixel: Pixel.Event
+        var errorToReport: Error?
+        var additionalParams: [String: String] = [:]
+
+        guard let error = error as? TerminationError else {
+            DailyPixel.fireDailyAndCount(pixel: .appDidTerminateWithUnhandledError, error: error)
+            fatalError("Unhandled error: \(error)")
+        }
+
+        defer {
+            DailyPixel.fireDailyAndCount(pixel: pixel, error: errorToReport, withAdditionalParameters: additionalParams)
+            switch mode {
+            case .immediately(let message):
+                fatalError(message)
+            case .afterAlert(let reason):
+                alertAndTerminate(with: reason)
+            }
+        }
+
+        switch error {
+        case .database(let error):
+            additionalParams = [
+                PixelParameters.applicationState: "\(application.applicationState.rawValue)",
+                PixelParameters.dataAvailability: "\(application.isProtectedDataAvailable)"
+            ]
+            switch error {
+            case .container(let error):
+                pixel = .dbContainerInitializationError
+                errorToReport = error
+                mode = .immediately(message: "DB container init failed: \(error.localizedDescription)")
+            case .other(let error):
+                pixel = .dbInitializationError
+                errorToReport = error
+                mode = error.isDiskFull
+                    ? .afterAlert(reason: .insufficientDiskSpace)
+                    : .immediately(message: "DB init failed: \(error.localizedDescription)")
+            }
+        case .bookmarksDatabase(let error):
+            pixel = .bookmarksCouldNotLoadDatabase
+            errorToReport = error
+            mode = error.isDiskFull
+            ? .afterAlert(reason: .insufficientDiskSpace)
+            : .immediately(message: "Bookmarks DB init failed: \(error.localizedDescription)")
```

Comment:
Somehow this is weird to read, indentation is not consistent with what's in line 98, have you tried to only keep : in new line, instead of ? and :

Reply:
Also, shall we localize these strings? Doesn't have to be right now, could be added later.

Reply:
these are not user facing, but used in fatalError as message